### DISCUSSION TITLE
How to get ElementsCollection in the page by XPath command

### DISCUSSION THREAD

**Discussion by mourav2222** [selenide/selenide#2872] [Q&A] [answered]
I want to check if the size of `ElementsCollection` of command `Selenide.$$x("//input[@formcontrolname='username']").size()` is 0 or 1.

The function `size()` is depricated and don't work reliable. There is a reference to using alternative of `size()` function
https://github.com/selenide/selenide/wiki/do-not-use-getters-in-tests

Use: `$$.shouldHave(size(n))`

My quetion is, how i can get an array of `ElementsCollection` with command `Selenide.$$x("//input[@formcontrolname='username']").shouldHave(size(1))` and use in my code later. 

If i used with `shouldBe(size(1))` command, my test can fail in some cases (if array is empty) size is 0, the user is logged yet. 
But i don't want check condition like Assert in test i want only get array of `ElementsCollection`.

```java
ElementsCollection myresult_array = Selenide.$$x("//input[@formcontrolname='username']").shouldHave(size(1));
```

I used return value ElementsCollection:

```java
 @CheckReturnValue
    @Nonnull
    public static ElementsCollection $$x(String xpathExpression) {
        return WebDriverRunner.getSelenideDriver().$$x(xpathExpression);
    }
```

How can i do this?

Thanks :)

**Comment by asolntsev** [comment]
@mourav2222 Sorry, I am not sure I can understand your question correctly. 

If you want to check that collection has 0 or 1 element, it's simple:
```java
$$x("//input[@formcontrolname='username']").shouldHave(sizeLessThanOrEqual(1));
```

P.S. Indeed, method `size()` is NOT deprecated. It's _not recommended_ because IMHO it leads to a bad test design, but technically it works perfectly.

**Reply by mourav2222** [reply]
@asolntsev  you didn't understand my question. I ask otherwise.
I have a login page and i want to know, if a user already logged or not.
I can check, if a login page appears or not with 

`$$x("//input[@formcontrolname='userna`m`e']").shouldHave(sizeLessThanOrEqual(1));`


But i want NOT stop the test on this command like assert it makes. I want only check and get a result and execute other commands in the test. 
Result: not logged -> do in If
Result: logged -> do in else

if(logged) [
 
 // code for logged

} else {

 // code for not logged

}
 
Can you follow my explanation?

Thanks...

**Reply by asolntsev** [reply, marked as answer]
Sure.  Then it's easy:

```java
if ($x("//input[@formcontrolname='username']").is(visible)) {
 ...
}
```

**Reply by mourav2222** [reply]
Many thanks! 
I didn't find why myself :)
